### PR TITLE
Add shortcuts for experimental QUERY verb

### PR DESCRIPTION
**Pull request by rabbiveesh**
### Summary
Add shortcuts for the new experimental QUERY http verb.

### Motivation
It was discussed on IRC. Bleeding edge features are a good look

### References
I brought up [this article](https://evertpot.com/get-request-bodies/).

### CODE REVIEW

**Comment by jberger** [comment]
I think for me the fact that mojo already supports any given method gives the flexibility to use bleeding edge features without having to codify them early. Until QUERY is standardized I'd be hesitant to add shortcut support. 👎 from me

**Comment by Grinnz** [comment]
This would need to be documented, and marked as experimental

**Comment by kraih** [comment]
Needs more work, but i like the feature. Been some time since we added something interesting from a bleeding edge spec.

**Comment by jberger** [comment]
> Needs more work, but i like the feature. Been some time since we added something interesting from a bleeding edge spec.

If you're comfortable with the inclusion, I'll defer to your judgement. You're certainly more on the pulse of these things

**Comment by rabbiveesh** [comment]
@kraih could you point me in the right direction? I'm not sure what you would like to be touched up

File: lib/Mojolicious/Routes/Route.pm [hunk 1]
```diff
@@ -521,6 +522,25 @@ L<Mojolicious::Guides::Routing> for more information.
...
+  my $route = $r->query('/:foo');
+  my $route = $r->query('/:foo' => sub ($c) {...});
+  my $route = $r->query('/:foo' => sub ($c) {...} => 'name');
+  my $route = $r->query('/:foo' => {foo => 'bar'} => sub ($c) {...});
+  my $route = $r->query('/:foo' => [foo => qr/\w+/] => sub ($c) {...});
+  my $route = $r->query('/:foo' => (agent => qr/Firefox/) => sub ($c) {...});
+
+Generate L<Mojolicious::Routes::Route> object matching only C<QUERY> requests, takes the same arguments as L</"any">
+(except for the HTTP methods to match, which are implied). See L<Mojolicious::Guides::Tutorial> and
+L<Mojolicious::Guides::Routing> for more information. Note that this HTTP verb is B<EXPERIMENTAL> and might change without warning.
```
Comment (kraih) [inline lib/Mojolicious/Routes/Route.pm:537]: Method, not verb.

File: lib/Mojolicious/Lite.pm [hunk 2]
```diff
@@ -293,6 +293,20 @@ L<Mojolicious::Guides::Tutorial> and L<Mojolicious::Guides::Routing> for more in
...
+  my $route = query '/:foo' => sub ($c) {...} => 'name';
+  my $route = query '/:foo' => {foo => 'bar'} => sub ($c) {...};
+  my $route = query '/:foo' => [foo => qr/\w+/] => sub ($c) {...};
+  my $route = query '/:foo' => (agent => qr/Firefox/) => sub ($c) {...};
+
+Generate route with L<Mojolicious::Routes::Route/"query">, matching only C<QUERY> requests. See
+L<Mojolicious::Guides::Tutorial> and L<Mojolicious::Guides::Routing> for more information.
+Note that this HTTP verb is B<EXPERIMENTAL> and might change without warning.
+
+This HTTP verb is part of L<a draft RFC|https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html>, which is still a work in progress.
```
Comment (kraih) [inline lib/Mojolicious/Lite.pm:308]: Line length.

File: lib/Mojo/UserAgent.pm [hunk 3]
```diff
@@ -901,6 +901,33 @@ callback.
...
+=head2 query
+
+  my $tx = $ua->query('example.com');
+  my $tx = $ua->query('http://example.com' => {Accept => '*/*'} => 'Content!');
+  my $tx = $ua->query('http://example.com' => {Accept => '*/*'} => form => {a => 'b'});
+  my $tx = $ua->query('http://example.com' => {Accept => '*/*'} => json => {a => 'b'});
+
+Perform blocking C<QUERY> request and return resulting L<Mojo::Transaction::HTTP> object, takes the same arguments as
+L<Mojo::UserAgent::Transactor/"tx"> (except for the C<QUERY> method, which is implied). You can also append a callback to
+perform requests non-blocking.
```
[the remark this answers is not in the document]
Comment (kraih) [reply lib/Mojo/UserAgent.pm:913]: Why no spec link here?

File: lib/Mojolicious/Lite.pm [hunk 4]
```diff
@@ -293,6 +293,20 @@ L<Mojolicious::Guides::Tutorial> and L<Mojolicious::Guides::Routing> for more in
...
+
+  my $route = query '/:foo' => sub ($c) {...};
+  my $route = query '/:foo' => sub ($c) {...} => 'name';
+  my $route = query '/:foo' => {foo => 'bar'} => sub ($c) {...};
+  my $route = query '/:foo' => [foo => qr/\w+/] => sub ($c) {...};
+  my $route = query '/:foo' => (agent => qr/Firefox/) => sub ($c) {...};
+
+Generate route with L<Mojolicious::Routes::Route/"query">, matching only C<QUERY> requests. See
+L<Mojolicious::Guides::Tutorial> and L<Mojolicious::Guides::Routing> for more information.
+Note that this HTTP verb is B<EXPERIMENTAL> and might change without warning.
```
[the remark this answers is not in the document]
Comment (kraih) [reply lib/Mojolicious/Lite.pm:306]: Line length.

File: lib/Mojolicious/Routes/Route.pm (hunk 1 above)
Comment (kraih) [reply lib/Mojolicious/Routes/Route.pm:537]: Line length.

File: lib/Mojolicious/Routes/Route.pm [hunk 5]
```diff
@@ -521,6 +522,25 @@ L<Mojolicious::Guides::Routing> for more information.
...
+
+  # Route with destination
+  $r->query('/user')->to('user#replace');
+  
+This HTTP verb is part of L<a draft RFC|https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html>, which is still a work in progress.
```
Comment (kraih) [inline lib/Mojolicious/Routes/Route.pm:542]: Line length.

File: lib/Test/Mojo.pm [hunk 6]
```diff
@@ -1016,6 +1017,19 @@ for the callback.
...
+=head2 query_ok
+
+  $t = $t->query_ok('http://example.com/foo');
+  $t = $t->query_ok('/foo');
+  $t = $t->query_ok('/foo' => {Accept => '*/*'} => 'Content!');
+  $t = $t->query_ok('/foo' => {Accept => '*/*'} => form => {a => 'b'});
+  $t = $t->query_ok('/foo' => {Accept => '*/*'} => json => {a => 'b'});
+
+Perform a C<QUERY> request and check for transport errors, takes the same arguments as L<Mojo::UserAgent/"query">, except
+for the callback. Note that this HTTP verb is B<EXPERIMENTAL> and might change without warning.
```
Comment (kraih) [inline lib/Test/Mojo.pm:1029]: Method, not verb.

File: lib/Test/Mojo.pm [hunk 7]
```diff
@@ -1016,6 +1017,19 @@ for the callback.
...
+
+This HTTP verb is part of L<a draft RFC|https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html>, which is still a work in progress.
```
Comment (kraih) [inline lib/Test/Mojo.pm:1031]: Line length.

File: t/mojolicious/lite_app.t [hunk 8]
```diff
@@ -552,6 +558,9 @@ $t->get_ok('/alterformat.html')->status_is(404)->header_is(Server => 'Mojoliciou
 $t->get_ok('/noformat')->status_is(200)->header_is(Server => 'Mojolicious (Perl)')->content_is('xml/noformat');
 $t->get_ok('/noformat.xml')->status_is(404)->header_is(Server => 'Mojolicious (Perl)')->content_is("Oops!\n");
 
+# Echo JSON via query request
+$t->query_ok('/echojson', json => {message => {an => [qw/object of sorts/]}})->json_is('/an' => [qw/object of sorts/]);
```
Comment (kraih) [inline t/mojolicious/lite_app.t:562]: You can also test with another method to make sure it's not just an `any` route.

File: lib/Mojo/UserAgent.pm (hunk 3 above)
[the remark this answers is not in the document]
Comment (kraih) [reply lib/Mojo/UserAgent.pm:913]: Method seems untested.

File: lib/Mojo/UserAgent.pm [hunk 9]
```diff
@@ -901,6 +901,33 @@ callback.
...
+
+  $ua->query('http://example.com' => json => {a => 'b'} => sub ($ua, $tx) { say $tx->result->body });
+  Mojo::IOLoop->start unless Mojo::IOLoop->is_running;
+
+=head2 query_p
+
+  my $promise = $ua->query_p('http://example.com');
+
+Same as L</"query">, but performs all requests non-blocking and returns a L<Mojo::Promise> object instead of accepting a
+callback.
```
[the remark this answers is not in the document]
Comment (kraih) [reply lib/Mojo/UserAgent.pm:923]: Also seems untested.

**Review by kraih — CHANGES_REQUESTED** [review]

**Comment by kraih** [comment]
PR seems abandoned. Given our growing review backlog, i'm closing this to make room for other PRs.

**Comment by rabbiveesh** [comment]
That's fair, I'll reopen it with the changes when I get time

**Pull request closed without merging**